src/Bonsai.ML.Visualizers/KinematicComponentVisualizer.cs

```diff
@@ -0,0 +1,157 @@
+        }
+
+        private PropertyInfo stateComponentProperty;
+
+        private int selectedIndex = 0;
+
+        /// <summary>
+        /// The selected index of the state component to be visualized
+        /// </summary>
+        public int SelectedIndex { get => selectedIndex; set => selectedIndex = value; }
+
+        /// <summary>
+        /// Size of the window when loaded
+        /// </summary>
+        public Size Size { get; set; }
+
+        /// <summary>
+        /// Capacity or length of time shown along the x axis of the plot during automatic updating
+        /// </summary>
+        public int Capacity { get; set; }
+
+        DateTime? _startTime;
+
+        TimeSeriesOxyPlotBase Plot;
+
+        public override void Load(IServiceProvider provider)
+        {
+            var stateComponents = GetStateComponents();
+            stateComponentProperty = typeof(KinematicComponent).GetProperty(stateComponents[selectedIndex]);
+
+            Plot = new TimeSeriesOxyPlotBase(
+                _lineSeriesName: "Mean",
+                _areaSeriesName: "Variance",
+                _dataSource: stateComponents,
+                _selectedIndex: selectedIndex
+            )
+            {
+                Size = Size,
+                Capacity = Capacity,
+                Dock = DockStyle.Fill,
+                StartTime = DateTime.Now
+            };
+
+            Plot.ResetSeries();
+
+            Plot.ComboBoxValueChanged += ComponentChanged;
+
+            var visualizerService = (IDialogTypeVisualizerService)provider.GetService(typeof(IDialogTypeVisualizerService));
+            if (visualizerService != null)
+            {
+                visualizerService.AddControl(Plot);
+            }
+        }
+
+        public override void Show(object value)
+        {
+            if (!_startTime.HasValue)
+            {
+                _startTime = DateTime.Now;
+                Plot.StartTime = _startTime.Value;
+                Plot.ResetSeries();
+            }
+
+            KinematicComponent kinematicComponent = (KinematicComponent)value;
+            StateComponent stateComponent = (StateComponent)stateComponentProperty.GetValue(kinematicComponent);
+            double mean = stateComponent.Mean;
+            double variance = stateComponent.Variance;
+
+            var time = DateTime.Now;
+
+            Plot.AddToLineSeries(
+                time: time,
+                mean: mean
+            );
+
+            Plot.AddToAreaSeries(
+                time: time,
+                mean: mean,
+                variance: variance
+            );
+
+            // var maxTime = Math.Ceiling(time);
+            // var minTime = time.AddSeconds(-Capacity);
+            // var minTime = time - _startTime;
+
+            if (time - _startTime.Value > TimeSpan.FromSeconds(Capacity))
+            {
+                Plot.SetAxes(minTime: time.AddSeconds(-Capacity), maxTime: time);
+            }
+
+            Plot.Update();
```

Comment:
There is a new `BufferedVisualizer` abstract class in the `Bonsai.Design` package which can probably take care of a lot of this logic for you, we can chat briefly about this and see if it is immediately applicable, otherwise we can leave it for later.